### DISCUSSION TITLE
How does redisson evict bad nodes when failover happen due to master node dying?

### DISCUSSION THREAD

**Discussion by johansjoberg** [redisson/redisson#5241] [Q&A] [answered]
Hi, I am wondering how redisson evict bad nodes when failover happen when using the clustered config setup. We are running 6 nodes and 3 shards, with 3 master-slave pairs. We are using Elasticache on AWS, so if a master node dies then AWS will automatically failover the cluster slaves to become master and replace the dead node a sa new slave node.

We however had issues where we got failures to create new connections in some cases because it seems like it still used the ip address of the old dead node. I tried to read the code that checks the state of the cluster to detect master-slave changes but don't find how eviction is made.
https://github.com/redisson/redisson/blob/594434fa6c2cb6dd7431dae85c43491e38e169d7/redisson/src/main/java/org/redisson/cluster/ClusterConnectionManager.java

Where in the code base is this type of eviction handled? To me it seems that in `ClusterConnectionManager` one only handle changes if new node addresses are discovered, but not if old ones are removed.

**Comment by mrniko** [comment]
> Where in the code base is this type of eviction handled?

in org.redisson.cluster.ClusterConnectionManager#checkMasterNodesChange method

> because it seems like it still used the ip address of the old dead node

It may happen if cluster topology is still contains this ip.

**Comment by johansjoberg** [comment]
@mrniko Thanks for the quick answer. While investigating further I suspect we found the issue. I suspect we have issues with JVM dns caching where we do not set a ttl for this when starting the JVM, which causes the original node addresses to be the ones that are always used. So the fix would be to set `networkaddress.cache.ttl` when starting the application to avoid that the dns lookup is reused forever.

**Reply by mrniko** [reply]
Netty uses async dns resolver and not a JVM's.

**Reply by johansjoberg** [reply]
@mrniko Where can I find how this is setup in the code base? I am wondering what TTL is used and if TTL is configurable.

**Reply by mrniko** [reply]
Here it is https://github.com/redisson/redisson/blob/master/redisson/src/main/java/org/redisson/connection/SequentialDnsAddressResolverFactory.java

**Reply by johansjoberg** [reply]
@mrniko Thanks. Actually we now think that this was a Redis server issue where for some reason we had 2 old nodes, that were replaced after a maintenance update, still being part of the cluster for ~30 min after failover finished. We reached out to AWS support to see if they have any logs etc. to confirm this.

From our view DNS cache TTL should use server default of 15s (for AWS Elasticache) so we should be fine and it should not be a DNS cache issue.

@mrniko Thanks for your quick replies and for being so helpful. It is much appreciated!

**Comment by johansjoberg** [comment, marked as answer]
For posterity, this was most likely not a DNS cache issue, but instead a Elasticache/Redis issue where old nodes where for some reason part of the cluster longer than usual after a completed failover.

**Reply by mrniko** [reply]
Unfortunately, it may happen with any Redis vendor
